From f8e2f241c1597209539f6a3032d767962132fc07 Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Tue, 20 May 2014 14:49:42 -0400
Subject: [s390] crypto: fix aes, des ctr mode concurrency finding

Message-id: <1400597382-12008-1-git-send-email-brueckner@redhat.com>
Patchwork-id: 79986
O-Subject: [RHEL7.1 PATCH] [s390] crypto: s390 - fix aes, des ctr mode concurrency finding
Bugzilla: 1097686
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
Author: Harald Freudenberger <freude@linux.vnet.ibm.com>

    crypto: s390 - fix aes,des ctr mode concurrency finding

    An additional testcase found an issue with the last
    series of patches applied: the fallback solution may
    not save the iv value after operation. This very small
    fix just makes sure the iv is copied back to the
    walk/desc struct.

    Cc: <stable@vger.kernel.org> # 3.14+
    Signed-off-by: Harald Freudenberger <freude@linux.vnet.ibm.com>

Bugzilla
--------
BZ 1097686
https://bugzilla.redhat.com/show_bug.cgi?id=1097686

Upstream status of the patch
----------------------------
The patch has been posted for upstream integration:
  http://www.spinics.net/lists/linux-crypto/msg10953.html
  http://www.spinics.net/lists/linux-crypto/msg10962.html

Meanwhile, the patch was accepted and integrated into the crypto-2.6 tree:
  https://git.kernel.org/cgit/linux/kernel/git/herbert/crypto-2.6.git/log/

This patch will be requested for a z-stream update.

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7475637

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/s390/crypto/aes_s390.c b/arch/s390/crypto/aes_s390.c
index 2a245b5..fd104db 100644
--- a/arch/s390/crypto/aes_s390.c
+++ b/arch/s390/crypto/aes_s390.c
@@ -818,6 +818,9 @@ static int ctr_aes_crypt(struct blkcipher_desc *desc, long func,
   else
    memcpy(walk->iv, ctrptr, AES_BLOCK_SIZE);
   spin_unlock(&ctrblk_lock);
+ } else {
+  if (!nbytes)
+   memcpy(walk->iv, ctrptr, AES_BLOCK_SIZE);
  }
  /*
   * final block may be < AES_BLOCK_SIZE, copy only nbytes
diff --git a/arch/s390/crypto/des_s390.c b/arch/s390/crypto/des_s390.c
index 2d96e68..f2d6ccc 100644
--- a/arch/s390/crypto/des_s390.c
+++ b/arch/s390/crypto/des_s390.c
@@ -429,6 +429,9 @@ static int ctr_desall_crypt(struct blkcipher_desc *desc, long func,
   else
    memcpy(walk->iv, ctrptr, DES_BLOCK_SIZE);
   spin_unlock(&ctrblk_lock);
+ } else {
+  if (!nbytes)
+   memcpy(walk->iv, ctrptr, DES_BLOCK_SIZE);
  }
  /* final block may be < DES_BLOCK_SIZE, copy only nbytes */
  if (nbytes) {
-- 
1.7.1